File: src/test/java/ash/java/graphql/test/schemas/GenreQueryTest.java
package ash.java.graphql.test.schemas;


import ash.java.graphql.test.TestUtil;
import ash.kotlin.graphql.TmdbSchema;
import ash.kotlin.graphql.data.GenreDao;
import ash.kotlin.graphql.fields.FieldProducer;
import ash.kotlin.graphql.fields.GenreSchema;
import ash.kotlin.graphql.types.genre.GenreType;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class GenreQueryTest {

    private static Object resultObjectIdName;
    private static JsonObject resultJsonIdName;
    private static JsonObject resultJsonId;
    private static JsonObject resultJsonName;

    @BeforeClass
    public static void setupResults() {
        TmdbSchema schema = new TmdbSchema(mockFields());

        resultObjectIdName = schema.executeQuery("{genres{id name}}");
        resultJsonIdName = TestUtil.extractData(resultObjectIdName);

        Object resultObjectId = schema.executeQuery("{genres{id}}");
        resultJsonId = TestUtil.extractData(resultObjectId);

        Object resultObjectName = schema.executeQuery("{genres{name}}");
        resultJsonName = TestUtil.extractData(resultObjectName);
    }

    @Test
    public void correctQueryShouldNotReturnNull() {
        assertThat(resultObjectIdName).isNotNull();
    }

    @Test
    public void correctQueryShouldReturnCorrectNumberOfGenres() {
        assertThat(resultJsonIdName.get("genres").getAsJsonArray().size()).isEqualTo(3);
    }

    @Test
    public void correctQueryShouldReturnListWithCorrectValues() {
        assertThat(resultJsonIdName.get("genres").getAsJsonArray().get(0).getAsJsonObject().get("id"))
                .isEqualTo(new JsonPrimitive(28));

        assertThat(resultJsonIdName.get("genres").getAsJsonArray().get(0).getAsJsonObject().get("name"))
                .isEqualTo(new JsonPrimitive("Action"));
    }

    @Test
    public void correctIdQueryReturnsListWithJustIds() {
        assertThat(resultJsonId.get("genres").getAsJsonArray().get(0).getAsJsonObject().get("id"))
                .isEqualTo(new JsonPrimitive(28));

        assertThat(resultJsonId.get("genres").getAsJsonArray().get(0).getAsJsonObject().get("name"))
                .isNull();
    }

    @Test
    public void correctNameQueryReturnsListWithJustNames() {
        assertThat(resultJsonName.get("genres").getAsJsonArray().get(0).getAsJsonObject().get("id"))
                .isNull();

        assertThat(resultJsonName.get("genres").getAsJsonArray().get(0).getAsJsonObject().get("name"))
                .isEqualTo(new JsonPrimitive("Action"));
    }

    private static List<FieldProducer> mockFields() {
        List<GenreType> genres = Arrays.asList(
                new GenreType(28, "Action"),
                new GenreType(35, "Comedy"),
                new GenreType(12, "Fantasy"));

        GenreDao genreDao = mock(GenreDao.class);

        when(genreDao.getAllMovieGenres()).thenReturn(genres);

        List<FieldProducer> fieldProducers = new ArrayList<>();
        fieldProducers.add(new GenreSchema(genreDao));

        return fieldProducers;
    }
}

File: src/test/kotlin/ash/kotlin/grapql/test/GenreQueryTest.kt
package ash.kotlin.grapql.test

import ash.kotlin.graphql.TmdbSchema
import ash.kotlin.graphql.data.TmdbUtil
import ash.kotlin.graphql.fields.GenreFieldDefinition
import ash.kotlin.graphql.types.genre.GenreType
import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
import org.assertj.core.api.Assertions.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner.Silent::class)
class GenreQueryTest
{
  private val genreArray = "genres"

  private val idField = "id"

  private val nameField = "name"

  private lateinit var resultObjectIdName: Any
  private lateinit var resultJsonIdName: JsonObject
  private lateinit var resultJsonId: JsonObject
  private lateinit var resultJsonName: JsonObject

  @Mock
  private lateinit var tmdbutil: TmdbUtil

  @InjectMocks
  private lateinit var fieldDefinition: GenreFieldDefinition

  @Before
  fun setupResults()
  {
    setupMock()

    val schema = TmdbSchema(listOf(fieldDefinition))

    resultObjectIdName = schema.executeQuery("{genres{id name}}")
    resultJsonIdName = extractData(resultObjectIdName)

    val resultObjectId = schema.executeQuery("{genres{id}}")
    resultJsonId = extractData(resultObjectId)

    val resultObjectName = schema.executeQuery("{genres{name}}")
    resultJsonName = extractData(resultObjectName)
  }

  @Test
  fun correctQueryShouldNotReturnNull()
  {
    assertThat(resultObjectIdName).isNotNull()
  }

  @Test
  fun correctQueryShouldReturnCorrectNumberOfGenres()
  {
    assertThat(resultJsonIdName[genreArray].asJsonArray.size()).isEqualTo(3)
  }

  @Test
  fun correctQueryShouldReturnListWithCorrectValues()
  {
    assertThat(resultJsonIdName[genreArray].asJsonArray[0].asJsonObject[idField])
        .isEqualTo(JsonPrimitive(28))

    assertThat(resultJsonIdName[genreArray].asJsonArray[0].asJsonObject[nameField])
        .isEqualTo(JsonPrimitive("Action"))
  }

  @Test
  fun correctIdQueryReturnsListWithJustIds()
  {
    assertThat(resultJsonId[genreArray].asJsonArray[0].asJsonObject[idField])
        .isEqualTo(JsonPrimitive(28))

    assertThat(resultJsonId[genreArray].asJsonArray[0].asJsonObject[nameField])
        .isNull()
  }

  @Test
  fun correctNameQueryReturnsListWithJustNames()
  {
    assertThat(resultJsonName[genreArray].asJsonArray[0].asJsonObject[idField])
        .isNull()

    assertThat(resultJsonName[genreArray].asJsonArray[0].asJsonObject[nameField])
        .isEqualTo(JsonPrimitive("Action"))
  }

  private fun setupMock()
  {
    `when`(tmdbutil.getGenreList()).thenReturn(listOf(
        GenreType(28, "Action"),
        GenreType(35, "Comedy"),
        GenreType(12, "Fantasy"))
    )
  }
}